http-clients/src/main/java/org/triplea/http/client/web/socket/WebSocketConnection.java

```diff
@@ -95,53 +109,71 @@ void close() {
   }
 
   /**
-   * Initiates a websocket connection. Must be called before {@link #sendMessage(String)}. This
-   * method blocks until the connection has either successfully been established, or the connection
-   * failed.
+   * Initiates a non-blocking websocket connection.
    *
-   * @throws CouldNotConnect If the connection fails
+   * @param errorHandler Invoked if there is a failure to connect.
+   * @throws IllegalStateException Thrown if connection is already open (eg: connect called twice).
+   * @throws IllegalStateException Thrown if connection has been closed (ie: 'close()' was called)
    */
-  void connect() {
+  // Suppression: Ignoring future return values ignores any exceptions thrown by the code
+  // that completes the future. In this case we ignore the returned future from 'whenComplete'
+  // to avoid blocking. We would not expect the ping sender start or error handler to throw
+  // exceptions.
+  @SuppressWarnings("FutureReturnValueIgnored")
+  void connect(final Consumer<String> errorHandler) {
     Preconditions.checkState(!client.isOpen());
     Preconditions.checkState(!closed);
-    try {
-      client.connect();
-      pingSender.start();
-    } catch (final Exception e) {
-      pingSender.cancel();
-      throw new CouldNotConnect(serverUri);
-    }
+
+    connectAsync()
+        .whenComplete(
+            (connected, throwable) -> {
+              if (connected && throwable == null) {
+                pingSender.start();
+              } else {
+                errorHandler.accept("Failed to connect to: " + serverUri);
+              }
+            });
+  }
+
+  // Suppression: Ignoring future return values ignores any exceptions thrown by the code
+  // that completes the future. Connect blocking does not throw any exceptions, there are
+  // no exceptions to ignore, we can ignore the future returned by 'submit'.
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private CompletableFuture<Boolean> connectAsync() {
+    final CompletableFuture<Boolean> completableFuture = new CompletableFuture<>();
+    // execute the connection attempt
+    Executors.newSingleThreadExecutor()
+        .submit(
+            () -> {
+              boolean connected;
```

Comment:
I'm really curious why this variable can't be final.
I get that in most cases try + catch can't guarantee that a variable is only "set once", but in this case it should be safe?
Is this just a compiler oddity?

Reply:
Indeed, compiler oddity, the assignment in the catch block flags a compiler error when made final.